### ISSUE TITLE
selectable bug: isSelectable() returns false for selectable item

### ISSUE THREAD

**Issue by or-dvir** [mikepenz/FastAdapter#718]
## Details
- [3.2.8] Used library version
- [27.1.1 ] Used support library version
- [3.1.4 ] Used tooling / Android Studio version

## Checklist

- [V ] Searched for [similar issues](https://github.com/mikepenz/FastAdapter/issues)
- [ V] Checked out the [sample application](https://github.com/mikepenz/FastAdapter/tree/develop/app)
- [ V] Read the [README](https://github.com/mikepenz/FastAdapter/blob/develop/README.md)
- [ V] Checked out the [CHANGELOG](https://github.com/mikepenz/FastAdapter/releases)
- [ V] Checked out the [MIGRATION GUIDE](https://github.com/mikepenz/FastAdapter/blob/develop/MIGRATION.md)

## About this issue
i have a list of expandable items and i am implementing selection on long click and i believe i found a bug in the class `ActionModeHelper`.
as per past implementations, my pre long click listener looks like this:
```
.withOnPreLongClickListener{v, adapter, item, position ->
                    mActionModeHelper.onLongClick(this, position) != null
                }
```
i have done everything else needed but items were not getting selected on long click. while investigating i noticed something strange in the library class `ActionModeHelper`, specifically in the method `onLongClick` (which is called inside the `withOnPreLongClickListener`).
i have placed a breakpoint on the first line of the method: 
`if (mActionMode == null && mFastAdapter.getItem(position).isSelectable())`

the problem i found is this: `mFastAdapter.getItem(position)` has a field `mSelectable` which is `true`, however calling `isSelectable()` returns false which causes the item to not be selected.

![image](https://user-images.githubusercontent.com/17179542/44775819-36008480-ab7f-11e8-8ad0-41cd5c970f28.png)

![image](https://user-images.githubusercontent.com/17179542/44776595-1cf8d300-ab81-11e8-8de7-8ea847bcbbb8.png)

as soon as i override the method `isSelectable()` inside my item and make it return always `true`, the problem was fixed.
however, i dont understand why this had to be done because as mentioned above the field `mSelectable` is already set to `true` even without overriding this method.
furthermore, i have another adapter in my app where i implement selection in exactly the same way (**without** overriding `isSelectable()`) and it works perfectly fine.
however that adapter is not expandable, perhaps the problem is related to expandable items only?

**Comment by mikepenz** [comment]
If you look at the `AbstractItem` it is implemented correctly:
https://github.com/mikepenz/FastAdapter/blob/develop/library-core/src/main/java/com/mikepenz/fastadapter/items/AbstractItem.java#L115

Not sure where you have a field `mIsSelectable` but the field in the abstract item is called `mSelectable` 

Ensure you do not have a field with the same name in your class extending the abstract item

**Comment by or-dvir** [comment]
this is a typing error (which is now corrected)... sorry.
the field is indeed called `mSelectable` (as seen in the screen shot)

**Comment by mikepenz** [comment]
have you figured out anything related to this? 

have you debugged though it, which field is used, where it comes from and if all correct methods are implemented and set the correct field?

**Comment by or-dvir** [comment]
my description of the problem in the original post is as far as i got. is there anything else specific you'd like to know?

**Comment by mikepenz** [comment]
hmmm other than you potentially debugging it further and see where this occurs, I can place guesses. 
To further investigate this I would also need some demo app, some sample based on the sample app or so to allow me to debug this case

**Comment by or-dvir** [comment]
@mikepenz please see this small [sample app](https://github.com/or-dvir/FastAdapterDeselectBug/tree/master/app/src/main/java/com/sample/fastadapterdeselectbug)

note: a small change is needed because this sample app was made for https://github.com/mikepenz/FastAdapter/issues/719

change needed:
in the class `EpisodeAdapterItem`, remove (or comment out) the line `override fun isSelectable() = true`

once you do that, items will not be selectable

**Comment by mikepenz** [comment]
Ok this one is also quite easy to describe. and I should have identified that earlier but it was a bit hard without actually being able to debug. 

So the default behaviour (and this is the most common case as far as I can think of applications) is that the parent is only selectable if it has no children. Thus means as your items all have subItems that their parents will not be selectable. This is defined in the `AbstractExpandableItem` and `ModelAbstractExpandableItem` 

You can fine the code here: https://github.com/mikepenz/FastAdapter/blob/develop/library-extensions-expandable/src/main/java/com/mikepenz/fastadapter/expandable/items/ModelAbstractExpandableItem.java#L64
I even state there that this might not be true for all applications, like yours as we figured out :D 

So overriding the function `isSelectable()` and returning true is actually the absolute correct behaviour (if you do not need any other behaviour) 

Sorry again for the inconvenience. 

And I hope this helps you :)

**Comment by or-dvir** [comment]
thank you mike. 
if i may make a suggestion:
not all of us dive into the source code. some just read the readme file and take a look into the sample app and start using the library (not just this one, any library). think it would be beneficial if you add a small explanation in the read me file like "please note that the default behavior for expandable item is to not be selectable".
its good that you wrote the comment in the code, but as i said not everyone looks into the source code to that comment is easily missed

thanks again for taking the time looking into this even though at the end its just the way the library was designed

**Comment by mikepenz** [comment]
so I should remove more docs to help people start looking into source code, which is in my opinion the most important thing a developer should do :). always stay curious always try to understand how things works. 
I can't even tell you how many hours I already spent in the android source code just to figure out how specific things work down on the bottom :P 

the problem with it. people don't even read the readme either, they are not even looking in the sample code. so it would not make such a big difference. 

btw. the behaviour is:
"Item is not selectable if it has child items" 

yeah sometimes it's hard to identify things just by a description. :) having code to look through is always the fastest for me :)

**Comment by or-dvir** [comment]
haha... completly understand and agree with you to some degree, but my opinion is a little different:
you dont necessarily need to fully understand how things work in order to use the - you can drive a car without being a mechanic and when you learn how to drive (at least where im from) they dont teach you how to build/fix a car. they teach you how to operate one.
please dont misunderstand me. the things im saying here are **not** criticism, but are simply my point of view on these things.

of course it helps a lot to fully understand the inner workings of the things you use but i just dont think its absolutely mandatory. and if i go back to the car example for a second, i do expect people to know how to change a tire but i dont expect them to know exactly how an engine works and every little piece of the "driving system". and also, just in case they really dont know anything, there is a manual :)

i believe that if you create something for other people to use (whether its a physical product or a piece of software or a library for other developers) then you should provide as much instructions as possible on how to use it. take for example this [library](https://github.com/or-dvir/EasySettings) i made - the readme file is fully detailed and the sample app is full of explanations. this way any developer can simply start using it and not have to read and analyse the entire library. just take it and use it (and here is _how_).

during most of my life whenever i was at a new place (school, job, army service, etc) and didnt really know how things worked i was simply told "figure it out" and i really hated that. why couldnt they just tell me how im supposed to do these things theyre  asking me to do? this is probably why i see things the way im trying to describe here.

in conclusion, i am a strong believer in RTFM - Read the F*** Manual :)

i repeat: this is **not** criticism, but are simply my point of view.

**Comment by mikepenz** [comment]
haha don't worry :D I love discussing these things. it's definitely one of the great things to discuss and see other peoples opinions on things. I mean after all this is open source. and if I did not care about others I would make it closed :D but I really really enjoy getting feedback, having discussions, iterating, making issues, fixing issues, improving and making it awesome! :D 

related to your RTFM :D I am the RTFJD (:P javadoc) or RTFC (code :P) 

but after all there are different personalities there are different approaches to serve for all cases. :)

related to your story. I think for me it's like I grew up trying to understand things, fiddle around, temper around, digging into, trying to understand, and then learn from the basics to know how a specific thing works. so I always fiddle around when I use a component and the public methods are not self describing, to see if there is some code hidden, if I can take it apart or make it behave the way I want ;) 


so about improving the readme, or improving the description. 

Please feel free to open a PR :) I am really really always happy to get people contribute, as this makes a open source project, open source. :D


Thanks a lot for your continuous usage, and support :) and for taking the time to write these thoughts!

**Comment by or-dvir** [comment]
there are 2 types of people i guess when it comes to this topic (ive seen this a lot throughout the years): those who learn by diving in, and those who learn by reading the instructions and dig a little further when necessary.
i guess youre the first and im the second.

no one is saying you dont care about others! you do amazing work! keep it up!

**Comment by mikepenz** [comment]
there is a 3rd type though :D people who neither learn by diving in, nor do they read instructions or dig a little bitter, and also google is no option for those :D 
luckily this type isn't too common :D. 

don't worry :) 

and please. if you have ideas on how to make stuff better. just open a PR ;)

**Comment by or-dvir** [comment]
ill try. i have a couple of projects which are due soon so my full attention is on those

**Comment by mikepenz** [comment]
no worries :) and no stress :D
